### DISCUSSION TITLE
Apple pro raw image support?

### DISCUSSION THREAD

**Discussion by FrontierDK** [ImageMagick/ImageMagick#3060] [Help]
Hi all.

Does anyone know if support for Apple pro raw will be added?

Here are some example files:
```
http://mitip.dk/IMG_0017.DNG
http://mitip.dk/IMG_0018.DNG
```

**Comment by fmw42** [comment]
Your links are bad. I get 404 errors.  If this is a DNG format, then ImageMagick supports DNG. Have you tried opening it with ImageMagick?

**Comment by FrontierDK** [comment]
Hi there - please try again. I did try right after posting, and noticed that the web server didn't have any mime types for .DNG. It does now.

**Reply by fmw42** [reply]
I can download, but then I get:

`magick DNG:IMG_0017.DNG IMG_0017.PNG`
magick: Unsupported file format or not RAW file `IMG_0017.DNG' @ error/dng.c/ReadDNGImage/420.

using ImageMagick 7.0.10.53.

Photoshop cannot open it either. But Mac Preview does open it.

**Comment by FrontierDK** [comment]
Exactly. These are Apple pro raw images, taken with an iPhone 12 pro max. ImageMagick will open .DNG from any Samsung phone without any issues, but not these photos...any chance of having this format included?

**Reply by fmw42** [reply]
I will leave that to the IM developers. But do you have any reference to the format. They will need that.

**Comment by urban-warrior** [comment]
Try this command:
```
magick JPEG:IMG_0017.DNG IMG_0017.PNG
```

**Reply by FrontierDK** [reply]
Hi - and thank you, that actually worked with the commandline version.
However, when using the DLL version, it doesn't work when setting JPEG: in front of the file name. Do you know how to get the DLL version to read the file too? :)

**Comment by urban-warrior** [comment]
You need to use the JPEG coder for this image, not the DNG DLL.  If you instead call ReadImage(), it will recognize JPEG: as an explicit image format type and properly route the call to the JPEG coder rather than the DNG coder.

**Comment by snibgo** [comment]
> These are Apple pro raw images, taken with an iPhone 12 pro max.

I don't think so. They seem to be JPEG files with the wrong extension.

**Reply by FrontierDK** [reply]
You're right, it seems iPhone / iOS is converting the files to jpeg when shared with another app. I'm making a HTML upload form, will post real DNG files soon.

**Reply by FrontierDK** [reply]
I have now uploaded the real .DNG file, and it seems ImageMagick does convert it - only, the colours are off. I guess I'll be playing with some .ICC profiles now...

Here is an example original file from iPhone 12 pro max, in case anyone want to play around.

Tip of the day: if you long press the download icon on iCloud.com - you get to select either to download a compressed version, or the uncompressed original.

```
http://mitip.dk/Test-original.DNG
http://mitip.dk/test-icloud.jpg
http://mitip.dk/Test-ImageMagick.jpg
```

**Comment by urban-warrior** [comment]
In addition to applying the right color transform, you might get some improvement with the DNG defines: dng:use-camera-wb, dng:use_auto_wb, dng:no-auto-bright, or dng:output-color, e.g. `-define dng:use-camera-wb=true`.

**Reply by FrontierDK** [reply]
Hi again 👍 
Tried using the flags above, but it doesn't bring me close to the original (as seen in http://mitip.dk/test-icloud.jpg  ).

When I do a simple open "blabla.dng" and save "blabla.jpg" in GraphicsMill, I get a file which has 99% of the same colours to test-icloud.jpg - and can be seen here: http://mitip.dk/GraphicsMill.jpg

However, when using ImageMagick, it doesn't make a large difference no matter how many and how I specify .ICC files and/or colorspaces.

Does anyone know how to get ImageMagick to end up with correct colours for DNG files? Feel free to spread/use example photos given.

**Reply by FrontierDK** [reply]
I wonder if there's a bug or I'm doing something wrong...it doesn't matter if I dng:use-camera-wb=true or set any value of dng:output-color - the image is for-for-bit the exact same, and has the same filesize. I'm using the 32bit version 7.0.10-53.

```
option explicit
dim image, result
set image=wscript.CreateObject("ImageMagickObject.MagickImage.1")
result=image.Convert("samsung_galaxy_note_20_ultra_03.dng", "-define", "dng:output-color=2", "2.jpg")
set image=nothing
```

**Comment by snibgo** [comment]
How did you make test-icloud.jpg and Test-ImageMagick.jpg? What were your commands?

Clearly, test-icloud.jpg has had a sharpening process that Test-ImageMagick.jpg has not had.

**Reply by FrontierDK** [reply]
A simple convert infile.dng outfile.jpg makes wrong colours.(very dark image).

Convert infile.dng -profile rgb.icc outfile.jpg produces dark images.

The images are really dark, and I wonder why GraphicsMill gets it spot on every time. So I guess some info isn't used in the conversion....

**Reply by fmw42** [reply]
Have you tried

Convert infile.dng -profile sRGB.icc outfile.jpg

and do you have your sRGB.icc profile in the same directory as your DNG file. If not, the specify the path to the sRGB.icc profile.

**Reply by FrontierDK** [reply]
Yup, tried just that (including full path to all files).

**Comment by snibgo** [comment]
Please link to rgb.icc. Is this a linear profile? If so, why are you converting to (or simply assigning) a linear profile?

Does the text result of "Convert infile.dng -verbose info:" show an embedded profile? If it doesn't, than your "-profile rgb.icc" will assign that profile. If the profile is not appropriate for your image, this is bad news.

What program are you using to view outfile.jpg? If that software ignores profiles. and a linear profile has been used, the image will look too dark.

**Reply by FrontierDK** [reply]
I'm a newbie when it comes to .ICC - but I still wonder why GraphicsMill will do a simple conversion, hit it spot on with the colours...not specifying any .ICC or any adjustment. I did an info -verbose on the original DNG file ( http://mitip.dk/Test-original.DNG ) and it is here: http://mitip.dk/Test-original.txt

When viewing .JPG files, I'm using the Windows Image viewer. http://mitip.dk/test-icloud.jpg and http://mitip.dk/GraphicsMill.jpg both appear as what can be seen on the iPhone.
```
http://mitip.dk/srgb.icc
http://mitip.dk/AdobeRGB1998.icc
http://mitip.dk/sRGB2014.icc
```

**Comment by fmw42** [comment]
Please show your GraphicsMill code and any default values being used.  What color management are you doing: profiles being attached, color management engines being used, color and tone corrections being used?

**Comment by fmw42** [comment]
See https://petapixel.com/2020/12/21/understanding-apple-proraw/ about new features apple added to DNG file format v1.6

**Comment by fmw42** [comment]
This is what I get from converting the DNG to JPG with no adjustments using IM 7

![Test-original_im7](https://user-images.githubusercontent.com/13458218/103467444-05f3be00-4d04-11eb-9c53-b2341dee3726.jpg)

And this is what I get from converting it in Photoshop with no adjustments

![Test-original_PS](https://user-images.githubusercontent.com/13458218/103467459-20c63280-4d04-11eb-890d-60520636a443.jpg)

**Reply by FrontierDK** [reply]
Both look very dark when compared to the iPhone.

What can be seen on iCloud.com: http://mitip.dk/test-icloud.jpg (original file from iCloud.com)
RAW DNG converted to JPEG, using GraphicsMill (old ActiveX version): http://mitip.dk/GraphicsMill.jpg

**Reply by fmw42** [reply]
Did you try applying an sRGB.icc profile in ImageMagick? That will be the brighter profile. I think there might be meta data that IM is not getting. See my link above about new features Apple has added to DNG specifications.

**Reply by FrontierDK** [reply]
Here is a very shortened version of a VBscript:
```
option explicit
dim img, result
set img = CreateObject("ImageMagickObject.MagickImage.1")
result = img.Convert("Test-original.DNG", "-profile", "sRGB.icc", "new.jpg")
set img=nothing
```
The GraphicsMill component, which does the job perfectly, is from 2004...almost 2 decades from the new features. So I have a feeling something else is at large...

**Reply by fmw42** [reply]
Perhaps they include other things such as white balancing by default

**Comment by urban-warrior** [comment]
We use libraw as our delegate library to read DNG images.  It returns a linear raw (demosaiced) image.  Its up to the photo editor to compose the right workflow to return the desired results.  You likely need a proper color profile transformation.  Note, color transforms require two color profiles, one is the source, one is the destination.  Your work flow only includes the one, sRGB.icc.

**Reply by fmw42** [reply]
The DNG file has no color profile according to IM identify -verbose.  Also perhaps libraw is missing some of the newer features of DNG added by Apple - ???

**Comment by FrontierDK** [comment]
I have tried specifying a number of different commands, both with one and 2 ICC profiles. I 100% admit that I'm a newbie to .ICC, I haven't been able to specify any combination of .ICC files, which enabled ImageMagick to end up with an image looking like what can be seen in GraphicsMill, the iPhone or on iCloud.com

I have a feeling I'm not the only one struggling with this issue, and that many will benefit from a solution :)

**Comment by snibgo** [comment]
The DNG file does not have a profile problem.

I get similar results to fmw42 from ImageMagick (using libraw) and from dcraw.

The files test-icloud.jpg and GraphicsMill.jpg have clearly had sharpening and increased saturation, and probably tonal adjustment.

IM will not do these things unless you tell it to.

**Comment by fmw42** [comment]
ImageMagick supports defines for added control, such as white balancing and adding profiles. See https://imagemagick.org/script/defines.php

**Comment by fmw42** [comment]
**I do not think the ImageMagick 7.0.10.54 defines for white balance are doing anything.** 

The verbose info lists:

  ` dng:wb.rb.levels: 2.144531 1.902832 1.000000 0.000000`

which should produce some white balancing if IM is dealing with them properly. But my testing with and without either white balancing method produces a compare value of 0 (0) no matter whether I use either define for white balancing or not.


Here are my commands:

```
magick DNG:Test-original.DNG Test-original_im7.jpg

magick DNG:Test-original.DNG -define dng:output-color=1 -define dng:use-camera-wb=true Test-original_defines.jpg

magick DNG:Test-original.DNG -define dng:output-color=1 -define dng:use-auto-wb=true Test-original_defines2.jpg

```
And the compare commands are:

```
magick compare -metric rmse Test-original_defines.jpg Test-original_im7.jpg null:
0 (0)

magick compare -metric rmse Test-original_defines2.jpg Test-original_im7.jpg null:
0 (0)

magick compare -metric rmse Test-original_defines2.jpg Test-original_defines.jpg null:
0 (0)

```

**Oddly, even the second white balance based on the image average shows no change (Test-original_defines2.jpg) relative to the no white balance (Test-original_im7.jpg)**


P.S. Also enabling auto-bright define also shows no difference when set to true or not used. 

magick DNG:Test-original.DNG -define dng:use-auto-bright=true Test-original_autobright.jpg

magick compare -metric rmse Test-original_autobright.jpg Test-original_im7.jpg null:
0 (0)

**So none of the DNG defines seem to be working.**

**Comment by snibgo** [comment]
@fmw42: Those defines affect the reading of the DNG file, so they must come before (not after) the input DNG.

However, white-balance processing occurs to mosaic data (CFA data), before demosaicing. But the image in Test-original.DNG isn't mosaic data. It is 12-bit JPEG, which has been white-balanced and demosaiced, so those defines would have no effect.

("ProRaw" files are not raw sensor data. They are the images made by the camera, possibly the results from composites.)

When used on a raw camera file that is mosaic data, those defines work fine.

**Reply by fmw42** [reply]
@snibgo 

Thanks for the tip about the defines. The defines section should be updated to indicate that they are for reading. 

But he supposedly reposted the raw DNG files. Here is the verbose info, which does indicate DNG format. I must admit, 

I know little about this format. But I will test the conversion again tomorrow using the defines before reading the input.

[Image:
  Filename: Test-original.DNG
  Base filename: Test-original.DNG
  Format: DNG (Digital Negative)
  Class: DirectClass
  Geometry: 4032x3024+0+0
  Units: Undefined
  Colorspace: sRGB
  Type: TrueColor
  Base type: Undefined
  Endianness: Undefined
  Depth: 16-bit
  Channel depth:
    Red: 16-bit
    Green: 16-bit
    Blue: 16-bit
](url)

**Comment by snibgo** [comment]
Yes, it is DNG format. But exiftool tells us it doesn't have a raw sensor image.

Exiftool on a DNG with a raw sensor image will have entries like:
```
Bits Per Sample                 : 14
Photometric Interpretation      : Color Filter Array
CFA Repeat Pattern Dim          : 2 2
CFA Pattern 2                   : 0 1 1 2
CFA Pattern                     : [Red,Green][Green,Blue]
Samples Per Pixel               : 1
```
... so each pixel has one 14-bit value that represents either red or green or blue.

But "exiftool Test-original.DNG" has this instead:
```
Bits Per Sample                 : 12 12 12
Compression                     : JPEG
Photometric Interpretation      : Linear Raw
Samples Per Pixel               : 3
```
... so each pixel has three 12-bit values.

On the subject of exiftool, we can extract the embedded JPEG "prettified" image:
```
exiftool -PreviewImage -b Test-original.DNG >preview.jpg
```
This looks the same as GraphicsMill.jpg, posted above. So I suspect that GraphicsMill is simply extracting the embedded JPEG.

preview.jpg is similar to test-icloud.jpg, but they are not identical, even apart from the cropping.

**Comment by urban-warrior** [comment]
ImageMagick utilizes the libraw delegate library to read DNG images.  It is based on `dcraw` and returns 16-bit images.  Download [dcraw](https://dechifro.org/dcraw/).  Produce a set of comand-line options that produces a satisfactory result.  Post your command-line options here.  We can emulate the results with the appropriate ImageMagick defines.

**Comment by snibgo** [comment]
dcraw can extract the embedded preview:
```
dcraw -e -T -O x.tiff Test-original.DNG
```
I'm not sure if this is what @FrontierDK wants.
